File: 程序员面试金典/面试题 02.06. 回文链表/Java/main3.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        if (head == null || head.next == null) return true;
        ListNode fast = head, slow = head;
        Deque<Integer> stack = new LinkedList<>();
        stack.push(head.val);
        while (fast.next != null && fast.next.next != null) {
            fast = fast.next.next;
            slow = slow.next;
            stack.push(slow.val);
        }

        if (fast.next == null) stack.pop();
        while (slow.next != null) {
            slow = slow.next;
            int tmp = stack.pop();
            if (tmp != slow.val) return false;
        }
        return true;
    }
}

File: 程序员面试金典/面试题 02.06. 回文链表/cpp/main3.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution
 {
 public:
     bool isPalindrome(ListNode *head)
     {
         if (!head || !head->next)
             return true;
         ListNode *slow = head, *fast = head;
         stack<int> st{{head->val}};
         while (fast->next && fast->next->next)
         {
             slow = slow->next;
             fast = fast->next->next;
             st.push(slow->val);
         }
         if (!fast->next)
             st.pop();
         while (slow->next)
         {
             slow = slow->next;
             int tmp = st.top();
             st.pop();
             if (tmp != slow->val)
                 return false;
         }
         return true;
     }
};